app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt

```diff
@@ -0,0 +1,395 @@
+          2
+        )
+      ).check(matches(withTagValue(equalTo(R.drawable.ic_onboarding_2))))
+    }
+  }
+
+  @Test
+  fun testOnboardingFragment_checkSlide2_skipButtonIsVisible() {
+    launch(OnboardingActivity::class.java).use {
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.skip_text_view)).check(matches(isDisplayed()))
+    }
+  }
+
+  @Test
+  fun testOnboardingFragment_checkSlide2_clickSkipButton_shiftsToLastSlide() {
+    launch(OnboardingActivity::class.java).use {
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.skip_text_view)).perform(click())
+      onView(
+        withIndex(
+          withId(R.id.slide_title_text_view),
+          1
+        )
+      ).check(matches(withText(getResources().getString(R.string.slide_3_title))))
+    }
+  }
+
+  @Test
+  fun testOnboardingFragment_checkSlide2_getStartedButtonIsNotVisible() {
+    launch(OnboardingActivity::class.java).use {
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.get_started_button)).check(matches(not(isDisplayed())))
+    }
+  }
+
+  @Test
+  fun testOnboardingFragment_checkSlide3Title_isCorrect() {
+    launch(OnboardingActivity::class.java).use {
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(
+        withIndex(
+          withId(R.id.slide_title_text_view),
+          2
+        )
+      ).check(matches(withText(getResources().getString(R.string.slide_3_title))))
+    }
+  }
+
+  @Test
+  fun testOnboardingFragment_checkSlide3Description_isCorrect() {
+    launch(OnboardingActivity::class.java).use {
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(
+        withIndex(
+          withId(R.id.slide_description_text_view),
+          2
+        )
+      ).check(matches(withText(getResources().getString(R.string.slide_3_description))))
+    }
+  }
+
+  @Test
+  fun testOnboardingFragment_checkSlide3Image_isCorrect() {
+    launch(OnboardingActivity::class.java).use {
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(
+        withIndex(
+          withId(R.id.slide_image_view),
+          2
```

Comment:
@BenHenning I am using one function `withIndex` which has been defined below. It helps to check the content for a specific item in ViewPager. But the issue is that the it will keep index only relative to the stack. For example: if there are 10 items in viewpager and you are going to last item of viewpager, which is position 9, the index for testing with this function will most probably be 2 or 3 considering that viewpager will have only that many loaded in its stack. So any suggestions on how can I fix this?

Reply:
This index thing seems a bit hacky, and the issue you describe seems difficult to solve robustly. Could we instead leverage tags. I'm thinking we tag each view pager view with information corresponding to the tab that view represents, then we can match based on the tag. It doesn't solve the needing to swipe bit, but we could conceivably implement a swipeToTab mechanism if we wanted to (though I think what you have now for swiping seems reasonable).

Reply:
@BenHenning In your suggestion, actually it would be similar to test the tag or text or anything for a view/viewgroup. If I am correct, the main reason behind your suggestion was to remove the dependency on index. So I finally removed the dependency on index and used below code which also, does not use tags.

```
@Test
  fun testOnboardingFragment_checkSlide3Image_isCorrect() {
    launch(OnboardingActivity::class.java).use {
      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
      onView(
        allOf(
          withId(R.id.slide_image_view),
          isCompletelyDisplayed()
        )
      ).check(matches(withTagValue(equalTo(R.drawable.ic_onboarding_3))))
    }
  }
```

you can see that I have used `isCompletelyDisplayed` which is need to make sure that the current view is actually visible on screen and is not in stack. The main issue that I am facing is that, because of animation and the time needed for swipe, the test case is failing. Which means that we need some sort of waiting mechanism. 

I took reference from `HomeActivityTest` and also https://stackoverflow.com/a/22563297/3689782 but none of these solutions seems to work. Also, if I put `SystemClock.sleep(1000L)` it does work perfectly. So can you suggest any suggestion on how can we fix this.

Also, in your HomeActivityTest, you have written code to wait for [30000L](https://github.com/oppia/oppia-android/blob/e3c6553e925c91c9c7860bd29f23c1e5de9d437a/app/src/sharedTest/java/org/oppia/app/home/HomeActivityTest.kt#L358) but when we run the app, it does work correctly but does not take 30 sec. So how exactly is it working.

In short, if we use `SystemClock.sleep(500)` it does work, its just that we need proper alternative of this.

Reply:
That HomeActivityTest mechanism waits up to 30 seconds for an action to occur and otherwise fails. It’s a bit friendlier to Espresso since it yields time for idling to properly occur. 

We should be disabling animations on emulators and devices when running tests. Is this actually an animation issue or a data loading/rendering issue? With animations disabled and onIdle satisfied, I’d expect this to pass unless we have coroutines somewhere blocking the UI (in which case this devolves to #89).

Reply:
Actually, animations are disabled but then also the sliding of viewpager and because of which these test cases are failing. Also, I did try your `HomeActivityTest` along with animation off but that did not work for me. I will try to investigate on this more today before confirming that whether it is needs #89 or not.

Reply:
@BenHenning I have created this https://gist.github.com/rt4914/17ef7c237b34fb4b5186080b14a78f16
which contains code that I have tried along with error and description.

Reply:
When you watch one of the failing tests run, what do you see? Is it actually swiping to the right tab, then hanging as the 30 second time waits to find the element but then fails despite the element actually being on the screen? Does it not swipe? Does it fail immediately after moving to the new tab?

Reply:
@BenHenning In the emulator it actually follows all the steps except waiting for 30 secs. It is working as if the 30 sec thing is not even working. (Same thing happens when i run HomeActivityTest). Also, all the steps in the emulator are actually working correctly but in the end test case fails with error: `The view was not visible 100%` because we are using `isCompletelyDisplayed()`.

Reply:
If it’s not waiting that means the condition that’s waiting 30 seconds is passing, which is presumably the withId. So it’s finding the element within 30 seconds, but something about the screen is causing isCompletelyDisplayed to fail. I’m not sure why that’s failing without investigating it in detail myself.

Reply:
@BenHenning I tried various things on this, mainly using the 30sec wait, but that was also not working. As a result I believe that you will need to investigate this issue.

Reply:
Two major points on how all these test cases have been fixed:

1. As per @BenHenning 's suggestion I replaced `swipe` with `ViewPagerActions.swipeToPage()` and that fixed some of the issues. `swipeToPage()` waits for everything to finish before going to the next step in test case.

2. Using `show-layout-bounds` from developer options in mobile, I was able to see some boxes which were actually meant to be `Visibility.GONE`, so that was also fixed, but changing the margin to padding.